File: android/src/main/kotlin/com/regula/plugin/facesdk/FlutterFaceApiPlugin.kt
package com.regula.plugin.facesdk

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.EventChannel.StreamHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding

const val channelID = "flutter_face_api"
val eventSinks = mutableMapOf<String, EventSink?>()

lateinit var args: List<Any?>
lateinit var binding: FlutterPluginBinding
val context: Context
    get() = binding.applicationContext

fun sendEvent(id: String, data: Any? = "") {
    eventSinks[id]?.let { Handler(Looper.getMainLooper()).post { it.success(data.toSendable()) } }
}

inline fun <reified T> argsNullable(index: Int) = when (val v = args[index]) {
    null -> null
    is Map<*, *> -> v.toJson() as T
    is List<*> -> v.toJson() as T
    else -> v as T
}

fun setupEventChannel(id: String) = EventChannel(binding.binaryMessenger, "$channelID/event/$id").setStreamHandler(object : StreamHandler {
    override fun onListen(arguments: Any?, events: EventSink) = events.let { eventSinks[id] = it }
    override fun onCancel(arguments: Any?) = Unit
})

class FlutterFaceApiPlugin : FlutterPlugin, MethodCallHandler {
    override fun onDetachedFromEngine(binding: FlutterPluginBinding) = Unit
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        args = call.arguments as List<*>
        methodCall(call.method) { data -> result.success(data.toSendable()) }
    }

    override fun onAttachedToEngine(flutterBinding: FlutterPluginBinding) {
        binding = flutterBinding
        for (event in arrayOf(
            cameraSwitchEvent,
            livenessNotificationEvent,
            videoEncoderCompletionEvent,
            onCustomButtonTappedEvent
        )) setupEventChannel(event)
        MethodChannel(binding.binaryMessenger, "$channelID/method").setMethodCallHandler(this)
    }
}

File: ios/flutter_face_api/src/FlutterFaceApiPlugin.swift
import Flutter

public class FlutterFaceApiPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channelID = "flutter_face_api"
        func setupEventChannel(_ eventId: String) {
            let channel = FlutterEventChannel(name: "\(channelID)/event/\(eventId)", binaryMessenger: registrar.messenger())
            channel.setStreamHandler(GenericStreamHandler(eventId))
        }
        setupEventChannel(cameraSwitchEvent);
        setupEventChannel(livenessNotificationEvent);
        setupEventChannel(videoEncoderCompletionEvent);
        setupEventChannel(onCustomButtonTappedEvent);
        
        let channel = FlutterMethodChannel(name: "\(channelID)/method", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(FlutterFaceApiPlugin(), channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        args = call.arguments as! [Any?]
        methodCall(call.method, { data in result(data.toSendable()) })
    }
} 

class GenericStreamHandler: NSObject, FlutterStreamHandler {
    private let eventId: String
    
    public init(_ eventId: String) {
        self.eventId = eventId
    }
    
    public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        eventSinks[eventId] = eventSink
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSinks[eventId] = nil
        return nil
    }
}

var eventSinks: [String: FlutterEventSink] = [:]
func sendEvent(_ event: String, _ data: Any? = nil) {
    DispatchQueue.main.async { eventSinks[event]?(data.toSendable()) }
}

func withPresenter(_ action: @escaping (UIViewController) -> Void) {
    DispatchQueue.main.async {
        let windows = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }.filter { $0.activationState == .foregroundActive || $0.activationState == .foregroundInactive }.flatMap { $0.windows }
        let candidates = windows.filter { $0.isKeyWindow } + windows.filter { !$0.isKeyWindow && !$0.isHidden && $0.alpha > 0 && $0.windowLevel == .normal }
        for window in candidates {
            guard var presenter = window.rootViewController else { continue }
            while let next = presenter.presentedViewController
                ?? (presenter as? UINavigationController)?.visibleViewController
                ?? (presenter as? UINavigationController)?.topViewController
                ?? (presenter as? UITabBarController)?.selectedViewController {
                presenter = next
            }
            guard presenter.viewIfLoaded?.window != nil else { continue }
            action(presenter)
            return
        }
        print("REGULA: Cannot present the UI: no presenter available.")
    }
}
